JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java

```diff
@@ -62,48 +75,147 @@ private void cleanupLeftovers(UUID currentId) {
+    /**
+     * Spawns a new Docker container with specified configurations.
+     *
+     * @param name Name of the container.
+     * @param startupScriptId Script to initialize the container with.
+     * @return The ContainerState of the newly created container.
+     */
+    public ContainerState initializeContainer(String name, StartupScriptId startupScriptId)
+            throws IOException {
+        if (startupScriptId == null || cachedContainers.isEmpty()
+                || !cachedContainers.containsKey(startupScriptId)) {
+            String containerId = createContainer(name);
+            return setupContainerWithScript(containerId, startupScriptId);
+        }
+        ContainerState containerState = cachedContainers.get(startupScriptId);
+        executor.submit(() -> initializeCachedContainer(startupScriptId));
+
+        client.renameContainerCmd(containerState.containerId()).withName(name).exec();
+        return containerState;
+    }
+
+    /**
+     * Initializes a new cached docker container with specified configurations.
+     *
+     * @param startupScriptId Script to initialize the container with.
+     */
+    private void initializeCachedContainer(StartupScriptId startupScriptId) {
+        String containerName = cachedContainerName();
+        String id = createContainer(containerName);
+        startContainer(id);
+
+        try {
+            ContainerState containerState = setupContainerWithScript(id, startupScriptId);
+            cachedContainers.put(startupScriptId, containerState);
+        } catch (IOException e) {
+            killContainerByName(containerName);
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * @param containerId The id of the container
+     * @param startupScriptId The startup script id of the session
+     * @return ContainerState of the spawned container.
+     * @throws IOException if an I/O error occurs
+     */
+    private ContainerState setupContainerWithScript(String containerId,
+            StartupScriptId startupScriptId) throws IOException {
+        startContainer(containerId);
+        PipedInputStream containerInput = new PipedInputStream();
+        BufferedWriter writer =
+                new BufferedWriter(new OutputStreamWriter(new PipedOutputStream(containerInput)));
+
+        InputStream containerOutput = attachToContainer(containerId, containerInput);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(containerOutput));
+
+        writer.write(Utils.sanitizeStartupScript(startupScriptsService.get(startupScriptId)));
+        writer.newLine();
+        writer.flush();
+
+        return new ContainerState(containerId, reader, writer);
+    }
+
+    /**
+     * Creates a new container
+     *
+     * @param containerId the ID of the container to start
+     */
+    private void startContainer(String containerId) {
+        if (!isContainerRunning(containerId)) {
```

Comment:
Log DEBUG when this method is called, the log should also contain if the if is true or false.